src/shared/Microsoft.Git.CredentialManager/Authentication/MicrosoftAuthentication.cs

```diff
@@ -45,10 +71,104 @@ public async Task<string> GetAccessTokenAsync(string authority, string clientId,
                 throw new Exception("Missing access token in response");
             }
 
-            return accessToken;
+            return new JsonWebToken(accessToken);
+        }
+
+        private async Task<JsonWebToken> GetAccessTokenInProcAsync(string authority, string clientId, Uri redirectUri, string[] scopes, string userName)
+        {
+            IPublicClientApplication app = await CreatePublicClientApplicationAsync(authority, clientId, redirectUri);
+
+            AuthenticationResult result = null;
+
+            // Try silent authentication first if we know about an existing user
+            if (!string.IsNullOrWhiteSpace(userName))
+            {
+                result = await GetAccessTokenSilentlyAsync(app, scopes, userName);
+            }
+
+            // If we failed to acquire an AT silently (either because we don't have an existing user, or the user's RT has expired)
+            // we need to prompt the user for credentials.
+            // Depending on the current platform and session type we try to show the most appropriate authentication interface.
+            if (result is null)
+            {
+#if NETFRAMEWORK
+                if (PlatformUtils.IsInteractiveSession())
+                {
+                    result = await app.AcquireTokenInteractive(scopes)
+                        .WithPrompt(Prompt.SelectAccount)
+                        .WithUseEmbeddedWebView(true)
+                        .ExecuteAsync();
+                }
+#elif NETSTANDARD
+                // MSAL requires the application redirect URI is a loopback address to use the System WebView
```

Comment:
Our `Microsoft.Git.CredentialManager` core library targets both `net461` and `netstandard2.0`.

We choose to also directly target `net461` so that MSAL can show WinForms-based UI for authentication (on Windows only). Although .NET Core 3.0 allows you to use WinForms (still Windows only) the MSAL library has not been updated to do this yet.

The code here tries to ensure that the best UI experience takes precedence on each runtime. Running on .NET Framework this means the WinForms 'embedded' UI that MSAL creates. For all other, non-specific platforms we do device-code flow*.

(*) Except remember we also have the native helper on macOS which takes an even higher precedence. Device code flow will help for scenarios when people are SSH'd into their machines.

Reply:
I think some extra comments (basically your description above) would be really helpful to add.

Reply:
Added!